File: python/brayns/core/geometry/capsule.py
# Copyright (c) 2015-2022 EPFL/Blue Brain Project
# All rights reserved. Do not distribute without permission.
#
# Responsible Author: adrien.fleury@epfl.ch
#
# This file is part of Brayns <https://github.com/BlueBrain/Brayns>
#
# This library is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License version 3.0 as published
# by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

from dataclasses import dataclass
from typing import Any

from brayns.utils import Vector3

from .geometry import Geometry


@dataclass
class Capsule(Geometry):
    """Capsule (cylinder with slope and half spheres at extremities).

    :param start_point: Cylinder base XYZ.
    :type start_point: Vector3
    :param start_radius: Cylinder and sphere base radius.
    :type start_radius: float
    :param end_point: Cylinder end XYZ.
    :type end_point: Vector3
    :param end_radius: Cylinder and sphere end radius.
    :type end_radius: float
    """

    start_point: Vector3
    start_radius: float
    end_point: Vector3
    end_radius: float

    @classmethod
    @property
    def method(cls) -> str:
        """JSON-RPC method to add capsules.

        :return: JSON-RPC method.
        :rtype: str
        """
        return 'add-capsules'

    def get_additional_properties(self) -> dict[str, Any]:
        """Low level API to serialize to JSON."""
        return {
            'p0': list(self.start_point),
            'r0': self.start_radius,
            'p1': list(self.end_point),
            'r1': self.end_radius,
        }

File: brayns/engine/geometry/types/Capsule.cpp
/* Copyright (c) 2015-2024, EPFL/Blue Brain Project
 * All rights reserved. Do not distribute without permission.
 * Responsible Author: Nadir Roman Guerrero <nadir.romanguerrero@epfl.ch>
 *
 * This file is part of Brayns <https://github.com/BlueBrain/Brayns>
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License version 3.0 as published
 * by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "Capsule.h"

#include <ospray/ospray_cpp/Data.h>

namespace
{
struct CapsuleParameters
{
    static inline const std::string type = "type";
    static inline const std::string basis = "basis";
    static inline const std::string position = "vertex.position_radius";
    static inline const std::string index = "index";
};
}

namespace brayns
{
Capsule CapsuleFactory::cylinder(const Vector3f &p0, const Vector3f &p1, const float radius) noexcept
{
    return Capsule{p0, radius, p1, radius};
}

Capsule CapsuleFactory::cone(const Vector3f &p0, const float r0, const Vector3f &p1, const float r1) noexcept
{
    return Capsule{p0, r0, p1, r1};
}

Capsule CapsuleFactory::sphere(const Vector3f &center, const float radius) noexcept
{
    return Capsule{center, radius, center + Vector3f(0.f, .01f, 0.f), radius};
}

Bounds GeometryTraits<Capsule>::computeBounds(const TransformMatrix &matrix, const Capsule &data)
{
    Vector3f p0Delta(data.r0);
    auto p0Min = data.p0 - p0Delta;
    auto p0Max = data.p0 + p0Delta;
    Vector3f p1Delta(data.r1);
    auto p1Min = data.p1 - p1Delta;
    auto p1Max = data.p1 + p1Delta;

    Bounds bounds;
    bounds.expand(matrix.transformPoint(p0Min));
    bounds.expand(matrix.transformPoint(p0Max));
    bounds.expand(matrix.transformPoint(p1Min));
    bounds.expand(matrix.transformPoint(p1Max));
    return bounds;
}

void GeometryTraits<Capsule>::updateData(ospray::cpp::Geometry &handle, std::vector<Capsule> &data)
{
    auto capsuleCount = data.size();

    std::vector<uint32_t> indexData;
    indexData.reserve(capsuleCount);
    for (uint32_t i = 0, index = 0; i < capsuleCount; ++i, index = index + 2)
    {
        indexData.push_back(index);
    }

    const auto type = OSPCurveType::OSP_ROUND;
    const auto basis = OSPCurveBasis::OSP_LINEAR;

    handle.setParam(CapsuleParameters::type, type);
    handle.setParam(CapsuleParameters::basis, basis);
    handle.setParam(CapsuleParameters::index, ospray::cpp::CopiedData(indexData));
    ospray::cpp::SharedData primitivesData(data.data(), OSPDataType::OSP_VEC4F, capsuleCount * 2);
    handle.setParam(CapsuleParameters::position, primitivesData);
}
}
